tests/unit/s2n_init_cleanup_leak_test.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+int allocations = 0;
+
+int s2n_mem_placeholder(void)
+{
+    return 0;
+}
+
+int s2n_mem_malloc(void **ptr, uint32_t size, uint32_t *given)
+{
+    *ptr = malloc(size);
+    if (*ptr) {
+        *given = size;
+        ++allocations;
+        return S2N_SUCCESS;
+    } else {
+        return S2N_FAILURE;
+    }
+}
+
+int s2n_mem_free(void *ptr, uint32_t size)
+{
+    (void)size;
+    if (ptr) {
+        free(ptr);
+        --allocations;
+    }
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
```

Comment:
I'm not convinced that this is necessary. We're investigating why valgrind didn't catch the leak, since it should be capable of it.

Reply:
Following the tenets of TDD, I thought it wise to first prove the problem exists and then solve it, which is what this test ultimately does.  The test is not harmful, and will protect against future issues in the init/cleanup cycle (until valgrind works correctly).  I have additional changes and tests for init and cleanup idempotency that I will submit after this, which will strengthen the s2n_init_test a bit.  If you do not want this test in the suite, I can remove it - it does fill a gap that seems to be missed by valgrind, which is quite perplexing!

Reply:
with https://github.com/aws/s2n-tls/pull/3511 valgrind should now catch "still reachable" leaks so I think the test is redundant. It was very useful for debugging though!

Reply:
I agree the test is redundant. We should be checking for leaks in all code paths, not just init/cleanup, which means we really need the valgrind fix instead of one single new test.

However, I could see the argument for leaving the test in to make sure valgrind is working properly, but we'd need clear comments explaining that if this test fails we need to investigate why valgrind isn't also failing. Like a valgrind sanity check.

But it looks like the memory leak here isn't really a huge deal, so I'm also more interested in this fix for preventing future errors (what if we add more important logic to s2n_cipher_suites_cleanup someday?) than for how it relates to memory.

Reply:
Based on the discussion and work on valgrind in #3511, I removed the additional test.